Add LoongArch support

jiegec opened the pull request:
- Comment out unsupported code
- Fill in operands


See https://github.com/capstone-engine/capstone/pull/2349

Rot127 committed:
Rebase refactored TableGen backends onto LLVM 18.

The MCInstDesc table changed. Bsides this only minor changes were done
and some additional code is emitted now for LLVM.

This commit is the combination of all previous Auto-Sync commits.
The list of commit messages follows:

-----------

Combination of all commits of the refactored tablegen backends.

These are the changes made for LLVM 16.

Refactor Capstone relevant TableGen Emitter backends.

This commit extracts the code which emits generated tables into two printer classes.
The Printer is called whenever actual code is written to a file.
There is the PrinterLLVM which emits tht code as before and
PrinterCapstone which is tailored to or needs (emitting C and generates
more info).

Additionally missing memory access properties were added to ARMs td
files.

Emit a single header for all files.

Captialize Target name for enums.

Add lay metric to emit enum value for Banked and system regs.

Malloc substr

Sort instructions in ascending order.

Free substr after use

Add vanished constrainsts

Fix `regInfoEmitEnums()` and indent

Fix `GenDisassemblerTables.inc#checkDecoderPredicate()`

Fix `TriCoreGenRegisterInfo.inc` | `PrinterCapstone::regInfoEmitRegClasses`

revert changes to NEON instructions

Add instructions with duplicate operands as Matchables.

Add memory load and store info

Correct memory access and out operand info

Set register lists again as read ops due to https://github.com/llvm/llvm-project/issues/62455

Make printAliasInstr and getMnemonic static.

Generate CS instruction enums from actual mnemonic. Not via the flawed AsmMatcher.

Fix typo in InstrInfoEmitter.cpp

Add deprecated QPX feature

Replace + and - with p and m

Add AssemblerPredicates to PPC

Generate RegEncodingTable

Define functions which are called by the Mapper as static.

Necessary because these functions are present in each arch'

Remove set_mem_access().

The cases where this is used to mark access to actual memory operands are
either very rare, or those are neon lane indicies.

Generate correct op type for absolute addresses.

Check for RegisterPointer operands first to prevent mis-categorization.

Add missing Operand types

Generate Instruction formats for PPC.

Add Paired Single instructions.

Partly revert 94e41ce23a7fd863a96288ec05b6c7202c3cfbf1 (introduces accidentially removed code.)

Set correct operand types for PS operands

Add memory read/write attributes

Add missing operand types

Add mayLoad and mayStore information.

Add documentation.

Handle special AArch64 operand

Replace C++ with C code.

Check for duplicate enum instr. names

Check for duplicate defintions of system registers.

Add note about missing target names.

Resolve templates in a single static method and add docs about it.

Revert printing target name in upper case.

Revert partially C++ syntax fixes in .td files.

They break the TemplateCOllector since it searches for exactly those references but can't find any'

Add all SubtargetFeatures to feature enum.

Not just the one used by CGIs.

Pass Decoder

Enable to check specific table fields to determine if reg enum must be emitted.

Allow to add namespace to type name/

Formatting

Rework emitting of tables.

The system operands are now emitted in reg, imm and aliass groups.
Also a bug was fixed which emitted incorrect code..

Check for rename IMPLICIT_IMM operand types

Pass DecodeComplete as pointer not as reference

Print undef when it needs to be printed.

Add namespace ids to all types and functions.

Rework C translation.

Pass MCOp as pointer not as ref

Add missing SysImm type

Fix syntax mistakes

Generate additonal sys immediates and op groups.

Handle edge case for printSVERegOp

Handle default arguments of template functions.

Add two missing op groups

Generate a static RecEncodingTable

Set enum values to encodings of the sys ops

Generate a single Enum value file for system operands.

Replace System operand groups with their operand types

Fix missing braces warning

Emit MCOperand validator.

Emit lookupByName functions for sys operands

Add namespaces for ARM.

Check for Target if default arguments of template functions are resolved.

auto-sync opcode & operand encoding info generation (#14)

* Added operand and opcode info generation

* Wrapped deprecated macro under an intellisense check

Basically intellisense fails, causing multiple errors in other files,

so when intellisense parses the code it will use the different version of the macro

* Fixed a small bug

Used double braces to prevent an old bug

Removed extra new line and fixed a bug regarding move semantics

jiegec committed:
Add LoongArch support

Rot127 committed:
Add script to check syntax and  compare LLVM upstream and our tables.

Rot127 committed:
Add CI jobs from main auto-sync branch again.

Rot127 committed:
Add table compare script to CI.

Rot127 committed:
Add missing commandline options for refactored backends.

Rot127 committed:
Add missing closing brackets in PPC def files.

Rot127 committed:
Update our def files to follow https://github.com/llvm/llvm-project/commit/4b43ef3e5c37459996ce0f53615881f436cb0e65

Rot127 committed:
Add a rebuild options and print more logs

Rot127 committed:
Blind fix for Github CI build

Rot127 committed:
Fix incorrect use of variable.

Rot127 committed:
Don't dump output of build in /dev/null

Rot127 committed:
Use Python provided cmake and Ninja for CI build and select gnu compiler 12

Rot127 committed:
Fix some incorrect generated LLVM code.

Rot127 committed:
Build debug llvm-tblgen in CI

Rot127 committed:
Add build instructions

Rot127 committed:
Separate generating of tables into different scripts so we can use Github action for branch checkout.

Rot127 committed:
Fix workflows

Rot127 committed:
Fix some incorrectly generated source code after rebase.

Rot127 committed:
Make gen scripts use the repository root dir.

Rot127 committed:
Remove syntax check, because it doesn't work that easy.

Rot127 committed:
Fix mismatches in generated C++ code.

Rot127 committed:
Fix mismatch in C++ Subtarget files

Rot127 committed:
Fix mismatch in generated C++ SystemOperands files.

Rot127 committed:
Fix no-return values warning from compiler.

Rot127 committed:
Remove instruction encoding information.

Rot127 committed:
Enable EmitMapTable to print C tables.

Rot127 committed:
Extends docs

Rot127 committed:
Add missing include guard to ignore list.

Rot127 committed:
Remove asserts

Rot127 committed:
Define the InstrTable as own type

Rot127 committed:
Add default argument for printSVERegOp

Rot127 committed:
Generate InstDecs tables with references to the OpInfo structs.

Rot127 committed:
Format code

Rot127 committed:
Fix: Don't return NULL for a struct

Rot127 committed:
Fix template function translation

Rot127 committed:
Add MatrixIndex_... to the OP_GROUP list

Rot127 committed:
Add AdrLabel and AdrpLable to OP_GROUPS

Rot127 committed:
Remove check for same name, different signature functions.

Rot127 committed:
Assign enum value to the raw_val member to prevent compiler warnings.

Rot127 commented on llvm/lib/Target/LoongArch/Disassembler/LoongArchDisassembler.cpp:
```diff
@@ -27,33 +27,33 @@ using namespace llvm;
+//                                          createLoongArchDisassembler);
+//   TargetRegistry::RegisterMCDisassembler(getTheLoongArch64Target(),
+//                                          createLoongArchDisassembler);
+// }
```
You don't need to comment out the unsupported code. With the auto-sync scripts you can choose to not copy them into the Capstone files later (the `Differ` step, see the [docs](https://github.com/capstone-engine/capstone/tree/next/suite/auto-sync/src/autosync/cpptranslator)).

In fact you shouldn't change any `cpp` or `h` files here (exceptions are additions to the `PrinterCapstone/LLVM` and fixes in the `td` files of LoongArch). I elaborate below

jiegec replied:
Thanks, I am more accustomed to generate everything and stick to them. If it is ok to keep out of sync for these files, I will proceed.

Rot127 commented on llvm/lib/Target/LoongArch/Disassembler/LoongArchDisassembler.cpp:
```diff
@@ -100,7 +100,7 @@ static DecodeStatus DecodeFCSRRegisterClass(MCInst &Inst, uint64_t RegNo,
 }
 
-template <unsigned N, int P = 0>
+template <unsigned N, int P>
```
These default arguments of template arguments should be handled by the [handleDefaultArg](https://github.com/Rot127/llvm-capstone/blob/f21eecd78d9b599d626b5567b9577d1631cd7cba/llvm/utils/TableGen/PrinterCapstone.cpp#L669-L703) function in `PrinterCapstone`.

jiegec replied:
I think I have tried this one before, but it did not support multiple template argument while one of them have default value. I will try again.

jiegec replied:
```
- #define DEFINE_decodeSImmOperand(N, S) \
+ #define DEFINE_decodeSImmOperand(N, unsigned S = 0) \
-       static DecodeStatus CONCAT(decodeSImmOperand, CONCAT(N, S))( \
+       static DecodeStatus CONCAT(decodeSImmOperand, \
+                                  CONCAT(N, unsigned S = 0))( \
                MCInst * Inst, uint64_t Imm, int64_t Address, \
                const void *Decoder) \
        { \
                MCOperand_CreateImm0(Inst, (SignExtend64((Imm << S), N + S))); \
                return MCDisassembler_Success; \
        }
```

jiegec replied:
Oh, I see what you mean. Fixing.

jiegec replied:
The missing feature is implemented in the last commit.

Rot127 commented on llvm/lib/Target/LoongArch/Disassembler/LoongArchDisassembler.cpp:
```diff
@@ -122,7 +122,7 @@ static DecodeStatus decodeSImmOperand(MCInst &Inst, uint64_t Imm,
 #include "LoongArchGenDisassemblerTables.inc"
 
-DecodeStatus LoongArchDisassembler::getInstruction(MCInst &MI, uint64_t &Size,
+DecodeStatus LoongArchDisassembler::LoongArch_LLVM_getInstruction(MCInst &MI, uint64_t &Size,
```
Renaming is not necessary. `LoongArchDisassembler::getInstruction` is translated to `LoongArchDisassembler_getInstruction` with the auto-sync scripts.

jiegec replied:
It somehow did not translate this for me.

Rot127 replied:
Ah, yes. So the reason is that every arch has this `getInstruction` function. Hence we can not expose it publicly. Because we would get linking conflicts.
Instead you can add the following function by hand to the translated `LoongArchDisassembler.c` file:

```c
DecodeStatus LoongArch_LLVM_getInstruction(csh handle, const uint8_t *Bytes,
				     size_t BytesLen, MCInst *MI,
				     uint16_t *Size, uint64_t Address,
				     void *Info)
{
	return getInstruction(handle, Bytes, BytesLen, MI, Size, Address, Info);
}
```

This function should be declared in the `LoongArchLinkage.h` header file. Same things need to be done for `getRegisterName` and `printInst`. You can check the PPC module as example.

Rot127 commented on llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:
```diff
@@ -209,7 +209,7 @@ def uimm8 : Operand<GRLenVT>, ImmLeaf<GRLenVT, [{return isUInt<8>(Imm);}]> {
 class UImm12Operand : Operand<GRLenVT>,
                       ImmLeaf <GRLenVT, [{return isUInt<12>(Imm);}]> {
-  let DecoderMethod = "decodeUImmOperand<12>";
+  let DecoderMethod = "decodeUImmOperand<12, 0>";
```
See above

Rot127 commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchInstPrinter.cpp:
```diff
@@ -56,27 +56,40 @@ void LoongArchInstPrinter::printInst(const MCInst *MI, uint64_t Address,
-  O << '$' << getRegisterName(Reg);
+static void LoongArchInstPrinter::printRegName(raw_ostream &O, MCRegister Reg) const {
+  // always use register alias name to avoid variable arguments
+  O << '$' << getRegisterName(Reg, LoongArch::RegAliasName);
```
This change is not necessary either. You can do it later in the C translated version of the file and save the change. Whenever the LoongArch module is updated, the change you made in the C code is automatically applied on the C file.

Rot127 commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchInstPrinter.cpp:
```diff
@@ -56,27 +56,40 @@ void LoongArchInstPrinter::printInst(const MCInst *MI, uint64_t Address,
+      MI->flat_insn->detail->loongarch.operands[MI->flat_insn->detail->loongarch.op_count].type = LOONGARCH_OP_REG;
+      MI->flat_insn->detail->loongarch.operands[MI->flat_insn->detail->loongarch.op_count].reg = MO.getReg();
+      MI->flat_insn->detail->loongarch.op_count++;
+    }
```
Please do not add the detail handling in the code. Every `printOperand` or similar function which handles operands, get a `add_cs_detail(...)` call added in the first line of the function. These functions should be implemented in `LoongArchMapping.c` and all the details are added there.

Rot127 commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchInstPrinter.cpp:
```diff
@@ -56,27 +56,40 @@ void LoongArchInstPrinter::printInst(const MCInst *MI, uint64_t Address,
-  MO.getExpr()->print(O, &MAI);
+  llvm_unreachable("Expressions are not supported.");
+  // assert(MO.isExpr() && "Unknown operand kind in printOperand");
+  // MO.getExpr()->print(O, &MAI);
```
The expression code can be later removed in the C translated version of the file.

Rot127 commented on llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchInstPrinter.h:
```diff
@@ -26,26 +26,26 @@ class LoongArchInstPrinter : public MCInstPrinter {
-  std::pair<const char *, uint64_t> getMnemonic(const MCInst *MI) override;
-  void printInstruction(const MCInst *MI, uint64_t Address,
+  // std::pair<const char *, uint64_t> getMnemonic(const MCInst *MI) override;
+  static void printInstruction(const MCInst *MI, uint64_t Address,
```
`static` should be changed later in the C code as well.

Rot127 commented on llvm/utils/TableGen/CodeGenTarget.cpp:
```diff
@@ -259,6 +259,7 @@ StringRef llvm::getEnumName(MVT::SimpleValueType T) {
   case MVT::Untyped:   return "MVT::Untyped";
   case MVT::funcref:   return "MVT::funcref";
   case MVT::externref: return "MVT::externref";
+  case MVT::INVALID_SIMPLE_VALUE_TYPE: return "MVT::INVALID_SIMPLE_VALUE_TYPE";
```
The invalid case should been handled. If not, please let me know the error you get.

jiegec replied:
```
INFO:root:Generating Mapping tables...
DEBUG:root:Generating CCS tables...
DEBUG:root:/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen --printerLang=CCS --gen-asm-matcher -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/include -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch/LoongArch.td
ILLEGAL VALUE TYPE!
UNREACHABLE executed at /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/utils/TableGen/CodeGenTarget.cpp:263!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen --printerLang=CCS --gen-asm-matcher -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/include -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch/LoongArch.td
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  llvm-tblgen 0x000056bedb92bfa0
1  llvm-tblgen 0x000056bedb92c39a
2  llvm-tblgen 0x000056bedb929e88
3  llvm-tblgen 0x000056bedb92b9cf
4  libc.so.6   0x00007d898dc77050
5  libc.so.6   0x00007d898dcc5e2c
6  libc.so.6   0x00007d898dc76fb2 gsignal + 18
7  libc.so.6   0x00007d898dc61472 abort + 211
```

Rot127 replied:
Ok, yes. This is actually the best way to fix it for us. Please add a comment like "capstone addition" to it. So we don't get confused when rebasing.

During looking into it I figured that the `bare_symbol` operand has no `OperandType` assigned. This procudes invalid operand details in the mapping tables like this:
```
...
  { CS_OP_INVALID, CS_AC_READ, { CS_DATA_TYPE_INVALID_SIMPLE_VALUE_TYPE, CS_DATA_TYPE_LAST } }, /* imm */
...
```
It should be an immediate I guess?

If yes, please apply this patch as well:
```diff
diff --git a/llvm/lib/Target/LoongArch/LoongArchInstrInfo.td b/llvm/lib/Target/LoongArch/LoongArchInstrInfo.td
index ae73a8ac74c7..af62cf9e74a0 100644
--- a/llvm/lib/Target/LoongArch/LoongArchInstrInfo.td
+++ b/llvm/lib/Target/LoongArch/LoongArchInstrInfo.td
@@ -450,6 +450,7 @@ def BareSymbol : AsmOperandClass {
 // A bare symbol used in "PseudoLA_*" instructions.
 def bare_symbol : Operand<GRLenVT> {
   let ParserMatchClass = BareSymbol;
+  let OperandType = "OPERAND_IMMEDIATE";
 }
 
 // Standalone (codegen-only) immleaf patterns.
```

jiegec replied:
Done

jiegec replied:
I also discovered that most immediate operands are not flagged as `OPERAND_IMMEDIATE`, so I add them in the following commit.

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -1784,7 +1784,7 @@ void PrinterCapstone::asmWriterEmitPrintAliasOp(
      << "         SStream *OS) {\n";
   if (PrintMethods.empty())
-    OS << "  llvm_unreachable(\"Unknown PrintMethod kind\");\n";
+    OS << "  assert(0 && \"Unknown PrintMethod kind\");\n";
```
`llvm_unreachable` gets automatically replaced by the translator.

jiegec replied:
The other uses of llvm_unreachable was replaced automatically, but this one was not.

Rot127 replied:
Sorry, where is is my mind. All good with this change.

Rot127 reviewed (changes requested):
Please, before you continue with this rebase it on top of the LLVM 18 branch from https://github.com/capstone-engine/llvm-capstone/pull/45. There are many bug fixes and it is based on LLVM-18.

Now to the changes:
- Before you proceed, please make sure you have read the architecture about the `auto-sync` updater. I would advice you to read this [blog entry](https://rizin.re/posts/auto-sync/) first and afterward check out the docs about: [the general architecture](https://github.com/capstone-engine/capstone/blob/next/docs/ARCHITECTURE.md), the [auto-sync updater](https://github.com/capstone-engine/capstone/blob/next/docs/AutoSync.md) and lastly the [cpp-translator](https://github.com/capstone-engine/capstone/tree/next/suite/auto-sync/src/autosync/cpptranslator). Last but not least I highly recommend to step with a debugger through a decoding of an ARM, PPC or AArch64 instruction. This way you can see in practice how it is done.
- Please do not make any changes to the LLVM files here. Exceptions are `TableGen` related files. But any target related source code should stay the same. The reason is, that it is a lot and very very annoying work to rebase our patched LLVM code onto the newest release. We need to have as little changes in the original LLVM source code as possible. This is why all our editing of the `LoongArchDisassembler` etc files, all happen in the auto-sync scripts and not here. Because the scritps automate a lot of them and keep the changes in this repo small.

You are one of the first users of the auto-sync updater. So I would love to improve the first steps for people like you. If you could tell me where you got confused I would like to know it, so I can make the "new-user" steps easier.

jiegec commented:
I have rebased on top of auto-sync-18 branch and dropped LoongArch-specific changes.

jiegec commented:
Using capstone `next` branch and llvm-capstone `auto-sync-18` branch:

```
INFO:root:Generating Disassembler tables...
DEBUG:root:Generating CCS tables...
DEBUG:root:/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen --printerLang=CCS --gen-disassembler -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/include -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch -o /home/jiegec/capstone/suite/auto-sync/build/llvm_c_inc/LoongArchGenDisassemblerTables.inc /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch/LoongArch.td
llvm-tblgen: Unknown command line argument '--printerLang=CCS'.  Try: '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen --help'
llvm-tblgen: Did you mean '--print-options=CCS'?
CRITICAL:root:Generation failed
Traceback (most recent call last):
  File "/home/jiegec/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 263, in <module>
    Updater.update()
  File "/home/jiegec/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 141, in update
    self.inc_generator.generate()
  File "/home/jiegec/capstone/suite/auto-sync/src/autosync/IncGenerator.py", line 101, in generate
    self.gen_incs()
  File "/home/jiegec/capstone/suite/auto-sync/src/autosync/IncGenerator.py", line 177, in gen_incs
    raise e
  File "/home/jiegec/capstone/suite/auto-sync/src/autosync/IncGenerator.py", line 171, in gen_incs
    subprocess.run(
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen', '--printerLang=CCS', '--gen-disassembler', '-I', '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/include', '-I', '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch', '-o', '/home/jiegec/capstone/suite/auto-sync/build/llvm_c_inc/LoongArchGenDisassemblerTables.inc', '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch/LoongArch.td']' returned non-zero exit status 1.
```

jiegec commented:
> Using capstone `next` branch and llvm-capstone `auto-sync-18` branch:
> 
> ```
> INFO:root:Generating Disassembler tables...
> DEBUG:root:Generating CCS tables...
> DEBUG:root:/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen --printerLang=CCS --gen-disassembler -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/include -I /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch -o /home/jiegec/capstone/suite/auto-sync/build/llvm_c_inc/LoongArchGenDisassemblerTables.inc /home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch/LoongArch.td
> llvm-tblgen: Unknown command line argument '--printerLang=CCS'.  Try: '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen --help'
> llvm-tblgen: Did you mean '--print-options=CCS'?
> CRITICAL:root:Generation failed
> Traceback (most recent call last):
>   File "/home/jiegec/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 263, in <module>
>     Updater.update()
>   File "/home/jiegec/capstone/suite/auto-sync/./src/autosync/ASUpdater.py", line 141, in update
>     self.inc_generator.generate()
>   File "/home/jiegec/capstone/suite/auto-sync/src/autosync/IncGenerator.py", line 101, in generate
>     self.gen_incs()
>   File "/home/jiegec/capstone/suite/auto-sync/src/autosync/IncGenerator.py", line 177, in gen_incs
>     raise e
>   File "/home/jiegec/capstone/suite/auto-sync/src/autosync/IncGenerator.py", line 171, in gen_incs
>     subprocess.run(
>   File "/usr/lib/python3.11/subprocess.py", line 571, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/build/bin/llvm-tblgen', '--printerLang=CCS', '--gen-disassembler', '-I', '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/include', '-I', '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch', '-o', '/home/jiegec/capstone/suite/auto-sync/build/llvm_c_inc/LoongArchGenDisassemblerTables.inc', '/home/jiegec/capstone/suite/auto-sync/vendor/llvm_root/llvm/lib/Target/LoongArch/LoongArch.td']' returned non-zero exit status 1.
> ```

Update: oh, I need to use auto-sync-18 from your repo, not capstone-engine.

jiegec committed:
Handle multiple template arguments in handleDefaultArg

jiegec committed:
Change RegDiffLists type to MCPhysReg

jiegec committed:
Avoid using llvm_unreachable

Rot127 commented:
Also please add LoongArch to the [CI](https://github.com/capstone-engine/llvm-capstone/tree/auto-sync/.github/workflows)

Rot127 committed:
Assign OPERAND_IMMEDIATE as OperandType of BareSymbol

jiegec committed:
Handle INVALID_SIMPLE_VALUE_TYPE in getEnumName

jiegec committed:
Set OperandType to OPERAND_IMMEDIATE for immediate operands

jiegec committed:
Emit formats enum and supplemental info for LoongArch

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2632,11 +2632,57 @@ std::string getArchSupplInfoPPC(StringRef const &TargetName,
+  // Compute memory access type
+  std::string MemoryAccess;
+  if (CGI->mayLoad && CGI->mayStore) {
+    MemoryAccess = "CS_AC_READ_WRTE";
```
Typo in `CS_AC_READ_WRTE` (missing `I` in `WRITE`)

jiegec replied:
It comes from the following enum:

```c++
typedef enum cs_ac_type {
	CS_AC_INVALID = 0,    ///< Uninitialized/invalid access type.
	CS_AC_READ = 1 << 0,  ///< Operand read from memory or register.
	CS_AC_WRITE = 1 << 1, ///< Operand write to memory or register.
	CS_AC_READ_WRTE =
		CS_AC_READ |
		CS_AC_WRITE, ///< Operand reads and writes from/to memory or register.
} cs_ac_type;
```

Rot127 replied:
Ah, well. This is a typo. Get's fixed here: https://github.com/capstone-engine/capstone/pull/2358

Rot127 committed:
Initialize DecoderComplete flag in generated decoder function.

Rot127 committed:
Check in patterns for memory operand properties.

- Determine memory operands by there membership of patterns.

Operands can be defined via patterns. The pattern has a type itself.
Patterns are commonly used to define memory operands
(as a collection of multiple other operands).

Here we check, if the operand is part of a pattern with type iPTR.
Type iPTR indicates a memory operand.

- Search for unspecified instruction patterns and extract memory op info from them.

AArch64 has the case of patterns which are not set in the CGI class.
Instead those patterns are free floating in the record keeper.

With this commit we first collect all such unassigned patterns and save them in a map
(indexed by instruction name they belong to).

When instruction operand details are generated, the operand type is searched in the pattern of the instruction.
If there is an operand with the same type in the pattern, which is also of a iPTR pattern, we treat it as memory operand.

Note that we can not compare the operands of the instruction and the pattern operands by name.
Because they simply are not named the same way.

This is a short coming in the td files and should be addressed in LLVM.
Because it is enormiously hard to get a 1 to 1 mapping from a pattern to an instruction
and its operands.

Rot127 committed:
Add memory access info as supplementary AArch64 info

Rot127 committed:
Fix regex pattern to not match operand names between ] and [

jiegec committed:
Rename CS_AC_READ_WRTE to CS_AC_READ_WRITE

Following https://github.com/capstone-engine/capstone/pull/2358

Rot127 commented:
Merged with https://github.com/capstone-engine/llvm-capstone/pull/45
